### PR TITLE
 Explicitly allow manual resume in EventDispatcher.nextEvent 

### LINKED ISSUE #154: Disallow resuming fiber in EventDispatcher.nextEvent if not explicitly asked
`EventDispatcher.nextEvent` accepts the flags, such as `Receive`, which will make `nextEvent` not throw an exception if the fiber is resumed with the data received - the default behavior is to allow resuming fiber for the reasons explicitly specified in the `nextEvent` call.

This behavior allows caller not to check the reason why the fiber was resumed - except in one case: the `nextEvent` will always allow explicit resume from the code with the positive resume code. This in turn makes user always to check that the resume reason is not explicit resume (with positive resume code). As this is allowed on relatively rare occasions, it seems that it might be better if the user would explicitly allow resuming the fiber (with another `NextEventFlag`).

Since this is breaking change (the behavior will change if you don't pass the flag everywhere), targeting next major.

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
EventDispatcher.nextEvent accepts the flags, such as Receive, which will
make nextEvent not throw an exception if the fiber is resumed with the
data received - the default behavior is to allow resuming fiber for the
reasons explicitly specified in the nextEvent call.

This behavior allows caller not to check the reason why the fiber was
resumed - except in one case: the nextEvent will always allow explicit
resume from the code with the positive resume code. This in turn makes
user always to check that the resume reason is not explicit resume (with
positive resume code). As this is allowed on relatively rare occasions,
it seems that it might be better if the user would explicitly allow
resuming the fiber (with another NextEventFlag).

This changes the logic so that the explicit `NextEventFlag.Resume` is
added, so the clients that want to be resumed should use it and they can
then drop the assertion that the fiber was not resumed manually from the
places they don't want this behaviour.

Fixes #154

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
Blocked on #153

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by nemanja-boric-sociomantic** [comment]
Oh, I forgot :police_car:

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the migration note.

**Comment by gavin-norman-sociomantic** [comment]
Rebase and unblock?

**Comment by nemanja-boric-sociomantic** [comment]
Rebased & unblocked.

File: relnotes/nextevent.migration.md
Comment (gavin-norman-sociomantic) [inline relnotes/nextevent.migration.md]: A bit of rewording for clarity (this is an important change):

"
The `EventDispatcher.nextEvent` method suspends the fiber and ensures that it is only resumed again for an expected reason. Previously, in addition to the resume reasons specified by the user when calling the method, manual resumption of the fiber with a positive resume code was always implicitly allowed. This meant that user code had to manually check for this case, if it was not desired.

Now, manual resumption of the fiber is disallowed by default, and `EventDispatcher.nextEvent` will by default treat this as a protocol error, if it occurs. Code that calls `nextEvent` should be updated as follows:
* If you wish to allow manual fiber resumption, pass the new `NextEventFlags.Resume` flag to `nextEvent`.
* If you wish to disallow manual fiber resumption, do nothing. For tidiness, you should remove any user code that checks for this occurrence.
"

File: src/swarm/neo/connection/RequestOnConnBase.d
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/connection/RequestOnConnBase.d]: Wait for explicit resume with a positive resume code

**Comment by nemanja-boric-sociomantic** [comment]
Rebased.

File: src/swarm/neo/request/RequestEventDispatcher.d
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/request/RequestEventDispatcher.d:583]: It occurs to me that these `waitingForX` methods should probably be combined, so as not to iterate over the array multiple times. Can be another PR.

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/RequestEventDispatcher.d:583]: Good idea, I'll follow-up with the next PR in series.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**
